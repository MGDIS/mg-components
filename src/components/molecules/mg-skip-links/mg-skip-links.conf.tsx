/**
 * type SkipLink
 */
export type SkipLink = {
  href: string;
  label: string;
};
