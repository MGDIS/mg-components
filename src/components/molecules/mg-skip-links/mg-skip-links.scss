.mg-skip-links {
  position: absolute;
  transform: translateY(-100%);
  opacity: 0;
  background: hsl(var(--color-light));

  &:focus-within {
    position: relative;
    transform: translateY(0);
    opacity: 1;
  }

  &__list {
    // Display
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    gap: 0.6rem;
    margin: 0;
    padding: 0.64rem 1.8rem;
    // Decoration
    list-style: none;
  }

  &__link {
    // Display
    display: inline-block;
    padding: 1.2rem;
    // Decoration
    color: hsl(var(--color-dark));

    &:hover {
      background: hsl(var(--color-dark), 0.1);
    }
  }
}
