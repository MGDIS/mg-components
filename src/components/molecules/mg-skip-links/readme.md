## Behavior

On the first tab, a banner with the link(s) is displayed by pushing the content down.

The number of links will be added according to the content of the page.

## Specs

### Sizing

The minimum height of the banner is 55px.

### Spacing

Items have an internal margin of 12px and are spaced 6px apart.

### Styling

- The links are in [@color-dark](./?path=/docs/style-colors--page).
- The background of the banner is in [@color-light](./?path=/docs/style-colors--page).

#### On hover

The background of the item takes the color [@color-dark](./?path=/docs/style-colors--page) at 10% opacity.

#### Focus taking

Native browser behavior is retained.

## Code example

You need to press <kbd>Tab</kbd> in the code example to see the component.

## Integration

The `mg-skip-links` must be integrated at the very top of your page to be the first focusable element.

## Implementation with a "#" based router

`mg-skip-links` uses native anchor behavior, but if your site/app uses a "#" link-based router, like AngularJS does, you'll need to use the "go-to-anchor" component event. This event returns its detail property to the target anchor, then you can apply the scrollTo anchor behavior:
- if the anchor is in the same window, with native javascript:

```js
const goToAnchor = (anchor) => {
    const top = document.querySelector(anchor).offsetTop;
    window.scrollTo(0, top);
}

Array.from(document.querySelector('mg-skip-links').shadowRoot.querySelectorAll('a')).forEach(anchor => {
    anchor.addEventListener('click', event => {
        event.preventDefault();
    })
})â€‹
```

- in case the anchor is in another window, with the [iframeRisizer's moveToAnchor method](https://github.com/davidjbradshaw/iframe-resizer/blob/master/docs/parent_page/methods.md#movetoanchoranchor)

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type         | Default     |
| -------- | --------- | ----------- | ------------ | ----------- |
| `links`  | --        | Skip links  | `SkipLink[]` | `undefined` |


## Events

| Event          | Description                       | Type                  |
| -------------- | --------------------------------- | --------------------- |
| `go-to-anchor` | Emited event when link is clicked | `CustomEvent<string>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
