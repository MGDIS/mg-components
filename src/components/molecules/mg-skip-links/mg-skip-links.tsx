import { Component, Element, Event, EventEmitter, h, Prop, Watch } from '@stencil/core';
import { SkipLink } from './mg-skip-links.conf';
import { initLocales } from '../../../locales';

@Component({
  tag: 'mg-skip-links',
  styleUrl: 'mg-skip-links.scss',
  shadow: true,
})
export class MgSkipLinks {
  /************
   * Internal *
   ************/

  private messages;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgSkipLinksElement;

  /**
   * Skip links
   */
  @Prop({ mutable: true }) links: SkipLink[];
  @Watch('links')
  validateLinks(links: SkipLink[]): void {
    if (
      links === undefined ||
      links.length === 0 ||
      links.some(link => link.href === undefined || !link.href.startsWith('#') || link.label === undefined || link.label.trim() === '')
    ) {
      throw new Error(`<mg-skip-links> prop "links": Cannot be empty and each link must contains an href starting with a "#" and a non empty label attributes.`);
    }
  }

  /**
   * Emited event when link is clicked
   */
  @Event({ eventName: 'go-to-anchor' }) goToAnchor: EventEmitter<string>;

  /***********
   * Methods *
   **********/

  /**
   * Handle link click and emit go-to-anchor event
   *
   * @param {MouseEvent} event mouse event
   * @returns {void}
   */
  private handleLinkCLick = (event: MouseEvent & { currentTarget: HTMLElement }): void => {
    this.goToAnchor.emit(event.currentTarget.getAttribute('href'));
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if props are well configured on init
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.validateLinks(this.links);
    this.messages = initLocales(this.element).messages;
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <nav class="mg-skip-links" role="navigation" aria-label={this.messages.skipLinks.navLabel}>
        <ul class="mg-skip-links__list">
          {this.links.map(link => (
            <li>
              <a class="mg-skip-links__link" href={link.href} onClick={this.handleLinkCLick}>
                {link.label}
              </a>
            </li>
          ))}
        </ul>
      </nav>
    );
  }
}
