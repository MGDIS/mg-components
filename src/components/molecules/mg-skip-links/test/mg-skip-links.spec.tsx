import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgSkipLinks } from '../mg-skip-links';

const getPage = args =>
  newSpecPage({
    components: [MgSkipLinks],
    template: () => <mg-skip-links {...args}></mg-skip-links>,
  });

describe('mg-skip-links', () => {
  test('with args %s', async () => {
    const { root } = await getPage({
      links: [
        { href: '#content', label: 'Content' },
        { href: '#menu', label: 'Menu' },
        { href: '#search', label: 'Search' },
        { href: '#footer', label: 'Footer' },
      ],
    });
    expect(root).toMatchSnapshot();
  });

  test.each([
    {},
    { links: [] },
    { links: [{ blu: 'blu', bli: 'bli' }] },
    { links: [{ href: 'nohash', Label: 'label' }] },
    { links: [{ href: '#hash', Label: ' ' }] },
    { links: [{ href: '#hash', Label: '' }] },
  ])('Should throw error with wrong link params: %s', async args => {
    expect.assertions(1);
    try {
      await getPage(args);
    } catch (err) {
      expect(err.message).toMatch('<mg-skip-links> prop "links": Cannot be empty and each link must contains an href starting with a "#" and a non empty label attributes.');
    }
  });

  test('should emit event when link is clicked', async () => {
    const page = await getPage({
      links: [
        { href: '#content', label: 'Content' },
        { href: '#menu', label: 'Menu' },
        { href: '#search', label: 'Search' },
        { href: '#footer', label: 'Footer' },
      ],
    });

    const spyGoToAnchor = jest.spyOn(page.rootInstance.goToAnchor, 'emit');

    const mgSkiplinks = page.doc.querySelector('mg-skip-links');
    const anchor = mgSkiplinks.shadowRoot.querySelector('a');
    const spyBlur = jest.spyOn(anchor, 'blur');
    anchor.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    await page.waitForChanges();

    expect(spyGoToAnchor).toHaveBeenCalledWith('#content');
    expect(spyBlur).toHaveBeenCalled();
  });
});
