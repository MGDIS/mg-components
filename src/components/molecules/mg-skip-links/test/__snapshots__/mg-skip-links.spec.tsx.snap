// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-skip-links with args %s 1`] = `
<mg-skip-links>
  <mock:shadow-root>
    <nav aria-label="Quick access" class="mg-skip-links" role="navigation">
      <ul class="mg-skip-links__list" role="list">
        <li>
          <a class="mg-skip-links__link" href="#content">
            Content
          </a>
        </li>
        <li>
          <a class="mg-skip-links__link" href="#menu">
            Menu
          </a>
        </li>
        <li>
          <a class="mg-skip-links__link" href="#search">
            Search
          </a>
        </li>
        <li>
          <a class="mg-skip-links__link" href="#footer">
            Footer
          </a>
        </li>
      </ul>
    </nav>
  </mock:shadow-root>
</mg-skip-links>
`;
